picard/coverart/providers/caa.py

```diff
@@ -68,34 +73,209 @@ def caa_url_fallback_list(desired_size, thumbnails):
+    """
+
+    LISTBOX_WIDTH = 150
+    LISTBOX_HEIGHT = 250
+
+    def __init__(self, parent=None):
+        super().__init__(parent=parent)
+        self.setFixedSize(QtCore.QSize(self.LISTBOX_WIDTH, self.LISTBOX_HEIGHT))
+        self.setSizePolicy(QtWidgets.QSizePolicy.Minimum, QtWidgets.QSizePolicy.Expanding)
+        self.setSortingEnabled(True)
+        self.setSelectionMode(QtWidgets.QAbstractItemView.ExtendedSelection)
+
+    def move_item(self, item, target_list):
+        """Move the specified item to another listbox."""
+        self.takeItem(self.row(item))
+        target_list.addItem(item)
+
+    def move_selected_items(self, target_list):
+        """Move the selected item to another listbox."""
+        for item in self.selectedItems():
+            self.move_item(item, target_list)
+
+    def move_all_items(self, target_list):
+        """Move all items to another listbox."""
+        while self.count():
+            self.move_item(self.item(0), target_list)
+
+    def move_all_items_from(self, source_lists):
+        """Move all items from another listbox."""
+        for source_list in source_lists:
+            while source_list.count():
+                item = source_list.item(0)
+                source_list.takeItem(source_list.row(item))
+                self.addItem(item)
+
+
 class CAATypesSelectorDialog(QtWidgets.QDialog):
-    _columns = 4
+    """Display dialog box to select the CAA image types to include and exclude from download and use.
 
-    def __init__(self, parent=None, types=None):
-        if types is None:
-            types = []
+    Keyword Arguments:
+        parent {[type]} -- Parent of the QDialog object being created (default: {None})
+        types_include {[string]} -- List of CAA image types to include (default: {None})
+        types_exclude {[string]} -- List of CAA image types to exclude (default: {None})
+    """
+
+    # Dictionary of standard image type names by translated image type title
+    #   key:    image type title displayed in list boxes (translated by i18n)
+    #   value:  standard image type name
+    #
+    # This is used to return the include and exclude lists containing standard image
+    # type names based on the titles displayed in the list boxes in the dialog.
+    image_types_by_display_title = {}
+
+    def __init__(self, parent=None, types_include=None, types_exclude=None):
         super().__init__(parent)
+        if types_include is None:
+            types_include = []
+        if types_exclude is None:
+            types_exclude = []
 
         self.setWindowTitle(_("Cover art types"))
-        self._items = {}
         self.layout = QtWidgets.QVBoxLayout(self)
 
+        # Populate display title / standard name dictionary
+        for caa_type in CAA_TYPES:
+            name = caa_type['name']
+            title = _(caa_type['title'])
+            self.image_types_by_display_title[title] = name
```

Comment:
Using translated string as key isn't a very good idea imho.
You should use the name which is meant for that (`caa_type['name']`)

Also `_()` will not work with caa types, because they come from another set of po files, you need to use `translate_caa_type()` (from coverart.utils)

Read about http://doc.qt.io/qt-5/qlistwidgetitem.html#setData and UserRole i think it is what you need here:
each item has caa type name as data, and you use `QListWidget.sortItems()`
It will prolly not work as expected because of QLocale stuff though (that's something we need to fix in Picard), don't bother with that for now.

Reply:
I am sure something like User Role will work. The easiest way is to just subclass it and add a field. See the toolbar customization options dialog in UI options for the relevant code.

Reply:
I struggled with that before but I think I have it now.  I'm storing the `caa_type['name']` in the list item ***UserRole*** role and the `caa_type['title']` as the display text.  I've also changed it to use `translate_caa_type()`.  It seems to be sorting the lists properly here for my testing, but everything is English so it's not much of a test.  I removed that extra dictionary.

This may sound strange, but one of my biggest problems is finding out which PyQt5 module a class or enumeration is in.  I had a heck of a time finding UserRole because I tried specifying it in PyQt5.Qt module instead of the PyQt5.QtCore.Qt module.  I guess I'll figure it out eventually.  :slightly_smiling_face: 

I'm just doing some final testing and clean-up and will upload shortly.